### PR TITLE
TheiaCoV and TheiaProk workflows standardization & organism default subworkflow

### LINKED ISSUE #290: Use the organism parameter workflow in all TheiaCoV workflows
:cool:

### :pushpin: Explain the Request

Implement the `wf_organism_parameters.wdl` in all TheiaCoV workflows

### :books: Context

Prevent mistakes based on incorrect user input

### PR DESCRIPTION
**Pull request by jrotieno**
Closes #290

## :hammer_and_wrench: Changes Being Made

We are making changes to variable names for standardization across TheiaCoV and TheiaProk workflows, and associated tasks. Specifically, we have made the following changes:

Previous   Name | New Name
-- | --
read1_raw | read1
read2_raw | read2
genome_size | genome_length
min_genome_size | min_genome_length
max_genome_size | max_genome_length
expected_genome_size | expected_genome_length
estimated_genome_size | estimated_genome_length
genome_size_output | genome_length_output
kmc_est_genome_size | kmc_est_genome_length
target_org | target_organism
percent_target_org | percent_target_organism

In addition, the `organism_parameters` logic workflow has been implemented that automatically sets organism-specific parameters for TheiaCoV organisms.

### Impacted Workflows/Tasks

task_dragonflye.wdl
task_shovill.wdl
task_nanoplot.wdl
task_screen.wdl
task_kraken2.wdl
task_broad_terra_tools.wdl
task_kmc.wdl
task_rasusa.wdl
wf_theiacov_illumina_pe.wdl
wf_theiacov_illumina_se.wdl
wf_theiacov_ont.wdl
wf_theiaprok_illumina_pe.wdl
wf_theiaprok_illumina_se.wdl
wf_theiaprok_ont.wdl
wf_read_QC_trim_ont.wdl
wf_read_QC_trim_pe.wdl
wf_read_QC_trim_se.wdl

all TheiaCoVs, Freyja_FASTQ, etc.

## :brain: Context and Rationale

As code base expands, inconsistencies arise that need standardization across old and new workflows. Such inconsistencies could include the use of a variable name that is less intuitive, e.g. genome length vs genome size, the use of two variable names in two workflows for the same input type, e.g. read1 vs read1_raw, etc.

For the organism parameter, this will reduce the number of static entries in the input JSONs so those only contain the variable parameters that change from week to week, more closely matching their intended purpose.

### Inputs

Change of input names, no new inputs

### Outputs

Change of output names, no new outputs

### Impacted Outputs

Not changed

## :test_tube: Testing

Picked random samples recently used for PHB 1.3.0 release validation.

### Locally

Not undertaken

### Terra

TheiaCoV Illumina PE: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/55978b2f-9e30-456c-b718-cf317e0ca339

TheiaCoV Illumina SE: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/1cad7bab-0b38-4f8b-b888-db1a4506ff0e

TheiaCoV ONT: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/05417f4f-c9f9-43f2-be06-c7d069e9b8b6

TheiaProk Illumina PE: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/e5828caf-e9d7-47f8-8f34-e1e18f6fa5ed

TheiaProk Illumina SE: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/ca2daa9c-8685-48de-abaa-d58c952daa4b

TheiaProk ONT: https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/964bc4a0-d0bc-4c38-99b4-6d25cd6ea6ec

## :microscope: Quality checks

Pull Request (PR) checklist:
- [X] Include a description of what is in this pull request in this message.
- [X] The workflow/task has been tested locally and on Terra
- [x] The CI/CD has been adjusted and tests are passing
- [X] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

File: workflows/utilities/wf_organism_parameters.wdl [hunk 1]
```diff
@@ -26,21 +29,25 @@ workflow organism_parameters {
...
     String sc2_nextclade_ref = "MN908947"
     String sc2_nextclade_ds_name = "sars-cov-2"
+    String sc2_pangolin_docker = "us-docker.pkg.dev/general-theiagen/staphb/pangolin:4.3.1-pdata-1.24"
     Int sc2_genome_len = 29903
     Int sc2_vadr_max_length = 30000
     String sc2_vadr_options = "--noseqnamemax --glsearch -s -r --nomisc --mkey sarscov2 --lowsim5seq 6 --lowsim3seq 6 --alt_fail lowscore,insertnn,deletinn --out_allfasta"
-
   }
-  if (organism == "MPXV") {
+  if (organism == "MPXV" || organism == "mpox" || organism == "monkeypox") {
```
Comment (kapsakcj) [inline workflows/utilities/wf_organism_parameters.wdl:49]: add `Mpox`, and any other common syntaxes

File: workflows/utilities/wf_organism_parameters.wdl [hunk 2]
```diff
@@ -63,7 +71,8 @@ workflow organism_parameters {
     String wnv_nextclade_ref = "NA"
     String wnv_nextclade_ds_name = "NA"
   }
-  if (organism == "flu") {
+  if (organism == "flu" || organism == "influenza") {
```
Comment (kapsakcj) [inline workflows/utilities/wf_organism_parameters.wdl:74]: add `Flu` and `Influenza`, others?

**Comment by kapsakcj** [comment]
I just merged `origin/main` back into this branch to resolve the conflict with TheiaMeta workflow WDL. See the above commit: [0f8b004](https://github.com/theiagen/public_health_bioinformatics/pull/310/commits/0f8b004948b36bb1afde2e524f83bb72ad9855b9)

File: tasks/assembly/task_shovill.wdl
Comment (kapsakcj) [inline tasks/assembly/task_shovill.wdl:1]: perhaps this is a good time to expose `cpu` and `memory` input parameters for shovill-se task

File: tasks/gene_typing/task_abricate.wdl [hunk 3]
```diff
@@ -60,13 +60,6 @@ task abricate_flu {
...
     String database = "insaflu"
-    String? nextclade_flu_h1n1_ha_tag
-    String? nextclade_flu_h1n1_na_tag
-    String? nextclade_flu_h3n2_ha_tag
-    String? nextclade_flu_h3n2_na_tag
-    String? nextclade_flu_vic_ha_tag
-    String? nextclade_flu_vic_na_tag
-    String? nextclade_flu_yam_tag
     Int minid = 70
     Int mincov =60
```
Comment (kapsakcj) [inline tasks/gene_typing/task_abricate.wdl:64]: formatting, add a space

File: tasks/taxon_id/task_kraken2.wdl
Comment (kapsakcj) [inline tasks/taxon_id/task_kraken2.wdl:1]: may be a good time to expose `memory` as an optional input

File: tasks/taxon_id/task_nextclade.wdl [hunk 4]
```diff
@@ -1,175 +1,175 @@
...
+        --output-json "~{basename}".nextclade.json \
+        --output-tsv  "~{basename}".nextclade.tsv \
+        --output-tree "~{basename}".nextclade.auspice.json \
+        --output-all=. \
+        "~{genome_fasta}"
+  >>>
+  runtime {
+    docker: "~{docker}"
+    memory: "4 GB"
+    cpu: 2
```
Comment (kapsakcj) [inline tasks/taxon_id/task_nextclade.wdl:45]: OK maybe I'm getting carried away here, as I've never need to increase resources for nextclade (it's lean and fast as is) but for uniformity-sake perhaps we also expose memory and cpus for the nextclade task

File: tasks/taxon_id/task_nextclade.wdl [hunk 5]
```diff
@@ -1,175 +1,175 @@
...
-              nc_clade = nc_clade + " (" + who_clade + ")"
-            if nc_clade == '':
-              nc_clade = 'NA'
-            Nextclade_Clade.write(nc_clade)
+  meta {
+    description: "Python and bash codeblocks for parsing the output files from Nextclade."
+  }
+  input {
+    File nextclade_tsv
+    String docker = "python:slim"
```
Comment (kapsakcj) [inline tasks/taxon_id/task_nextclade.wdl:67]: FYI/warning, we may want to consider switching to a static docker image. The `python:slim` docker image from dockerhub is updated frequently without warning: https://hub.docker.com/_/python/tags?page=1&name=slim

I haven't seen this fail ever, as it's some pretty vanilla python code but would be good to consider

File: workflows/theiacov/wf_theiacov_clearlabs.wdl [hunk 6]
```diff
@@ -12,14 +12,15 @@ import "../../tasks/species_typing/task_pangolin.wdl" as pangolin
...
 
 workflow theiacov_clearlabs {
   meta {
     description: "Reference-based consensus calling for viral amplicon ont sequencing data generated on the Clear Labs platform."
   }
   input {
     String samplename
-    File clear_lab_fastq
+    File read1
     String organism = "sars-cov-2"
```
Comment (kapsakcj) [inline workflows/theiacov/wf_theiacov_clearlabs.wdl:24]: I'm glad that this is kept as default, but most all other default input variables (specific to sars-cov-2) are now in the organism parameters subwf. Soooo nice

File: workflows/theiacov/wf_theiacov_illumina_pe.wdl [hunk 7]
```diff
@@ -157,96 +171,133 @@ workflow theiacov_illumina_pe {
...
+              reference_gff_file = reference_gff,
+              reference_genome = reference_genome,
+              genome_length = genome_length,
+              nextclade_ds_reference = nextclade_dataset_reference,
+              nextclade_ds_tag = nextclade_dataset_tag,
+              nextclade_ds_name = nextclade_dataset_name,     
+              vadr_max_length = vadr_max_length,
+              vadr_options = vadr_options,
+              primer_bed_file = primer_bed,
+              pangolin_docker_image = pangolin_docker_image,
```
Comment (kapsakcj) [inline workflows/theiacov/wf_theiacov_illumina_pe.wdl:191]: can we remove `pangolin_docker_image` from this input block (and additionally the input block below for `set_flu_ha_nextclade_values`?)

Looks like an optional input for organism param subwf so might be good to remove

Comment (kapsakcj) [reply workflows/theiacov/wf_theiacov_illumina_pe.wdl:191]: Ah, now I see the comment. disregard!

File: workflows/theiacov/wf_theiacov_ont.wdl [hunk 8]
```diff
@@ -272,6 +309,9 @@ workflow theiacov_ont {
       }
     }
   }
+  call versioning.version_capture{
```
Comment (kapsakcj) [inline workflows/theiacov/wf_theiacov_ont.wdl:312]: Any particular reason why this task was moved to the end? I suppose it does not really matter/affect things, I'm just curious why this was moved

**Comment by kapsakcj** [comment]
All the code changes LGTM. I had a few questions but regardless I can start testing in Terra early next week

File: workflows/theiacov/wf_theiacov_ont.wdl (hunk 8 above)
Comment (sage-wright) [reply workflows/theiacov/wf_theiacov_ont.wdl:312]: just to make it prettier and mimic the other workflows

**Comment by kapsakcj** [comment]
Just merged in `main` which includes Ines' PR240 on Kraken & ONT data as well as mine/James' PR275 on unaligned reads out of TheiaCov Illumina workflows 

I expect CI to fail and I'll update shortly.

@cimendes It would be good to double check my work on resolving conflicts regarding the kraken task & ONT readQC subworkflow. There were a lot of places where I changed `target_org` to `target_organism` in both WDL and bash variables and I hope I updated everything correctly 🤞

**Comment by cimendes** [comment]
@kapsakcj I've resolved the conflicts and will update hashes once all the tests terminate :)

**Comment by cimendes** [comment]
Functional and TheiaValidate validation performed in https://app.terra.bio/#workspaces/theiagen-validations/PHB_Validation_PR_310

File: tests/workflows/theiaprok/test_wf_theiaprok_illumina_pe.yml [hunk 9]
```diff
@@ -570,7 +570,7 @@
       md5sum: c8e7ae1ed0fff7b14731228db80c6b30
     - path: miniwdl_run/wdl/tasks/quality_control/read_filtering/task_fastp.wdl
       md5sum: 0faa3de4c2e4aa4cbbe95ae4376fe33a
-    - krakath: miniwdl_run/wdl/tasks/quality_control/basic_statistics/task_fastq_scan.wdl
```
Comment (sage-wright) [inline tests/workflows/theiaprok/test_wf_theiaprok_illumina_pe.yml]: what do you mean krakath isn't right

**Comment by kapsakcj** [comment]
TheiaEuk_Illumina_PE functional test here (launched after all of these commits were made): https://app.terra.bio/#workspaces/theiagen-validations/PHB_Validation_PR_310/job_history/79d3dc36-da4f-470a-af59-ccfd085ae5cc

waiting for it to finish, so need to review results

**Review by cimendes — APPROVED** [review]
This PR has been extensively validated in the workspace I've shared previously. Feeling very happy with the outcome of the validation exercise! Great job team!

**Comment by kapsakcj** [comment]
we cannot forget to update the documentation with this PR change

**Pull request merged**